axum/src/serve.rs

```diff
@@ -19,12 +18,59 @@ use hyper_util::{
     server::conn::auto::Builder,
 };
 use tokio::{
+    io::{AsyncRead, AsyncWrite},
     net::{TcpListener, TcpStream},
     sync::watch,
 };
 use tower::util::ServiceExt;
 use tower_service::Service;
 
+/// TODO
+pub trait Listener: Send + 'static {
+    /// The listener's IO type.
+    type Io: AsyncRead + AsyncWrite + Unpin + Send + 'static;
+
+    /// The listener's address type.
+    type Addr: Send;
+
+    /// Accept a new incoming connection to this listener
+    fn accept(&mut self) -> impl Future<Output = io::Result<(Self::Io, Self::Addr)>> + Send;
+
+    /// Returns the local address that this listener is bound to.
+    fn local_addr(&self) -> io::Result<Self::Addr>;
+}
```

Comment:
We could also use tokio-util's [`Listener`](https://docs.rs/tokio-util/latest/tokio_util/net/trait.Listener.html) trait however I don't think tokio-util should be in our public API. By writing our own we can also use `async fn accept` rather than `fn poll_accept`.